pkg/services/store/resolver/service.go

```diff
@@ -0,0 +1,122 @@
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/manager/registry"
+	"github.com/grafana/grafana/pkg/services/datasources"
+)
+
+// for testing
+var getNow = func() time.Time { return time.Now() }
+
+type ResolutionInfo struct {
+	OK        bool      `json:"ok"`
+	Key       string    `json:"key,omitempty"`  // GRN? UID?
+	Warning   string    `json:"kind,omitempty"` // old syntax?  (name>uid) references a renamed object?
+	Timestamp time.Time `json:"timestamp,omitempty"`
+}
+
+type ObjectReferenceResolver interface {
+	Resolve(ctx context.Context, ref *models.ObjectExternalReference) (ResolutionInfo, error)
+}
+
+func ProvideObjectReferenceResolver(ds datasources.DataSourceService, pluginRegistry registry.Service) ObjectReferenceResolver {
+	return &standardReferenceResolver{
+		pluginRegistry: pluginRegistry,
+		ds: dsCache{
+			ds:             ds,
+			pluginRegistry: pluginRegistry,
+		},
+	}
+}
+
+type standardReferenceResolver struct {
+	pluginRegistry registry.Service
+	ds             dsCache
+}
+
+func (r *standardReferenceResolver) Resolve(ctx context.Context, ref *models.ObjectExternalReference) (ResolutionInfo, error) {
+	if ref == nil {
+		return ResolutionInfo{
+			OK:        false,
+			Timestamp: getNow(),
+			Warning:   "invalid reference (nil)",
+		}, nil
+	}
+
+	switch ref.Kind {
+	case models.StandardKindDataSource:
+		return r.resolveDatasource(ctx, ref)
+
+	case models.ExternalEntityReferencePlugin:
+		return r.resolvePlugin(ctx, ref)
+
+	case models.ExternalEntityReferenceRuntime:
+		return ResolutionInfo{
+			OK:        false,
+			Timestamp: getNow(),
+			Warning:   "not implemented yet", // TODO, runtime registry?
+		}, nil
+	}
+
+	return ResolutionInfo{
+		OK:        false,
+		Timestamp: getNow(),
+		Warning:   "resolution not yet implemented",
+	}, nil
+}
+
+func (r *standardReferenceResolver) resolveDatasource(ctx context.Context, ref *models.ObjectExternalReference) (ResolutionInfo, error) {
+	ds, err := r.ds.getDS(ctx, ref.UID)
+	if err != nil || ds == nil || ds.UID == "" {
+		return ResolutionInfo{
+			OK:        false,
+			Timestamp: r.ds.timestamp,
+		}, err
+	}
+
+	res := ResolutionInfo{
+		OK:        true,
+		Timestamp: r.ds.timestamp,
+		Key:       ds.UID, // TODO!
+	}
+	if !ds.PluginExists {
+		res.OK = false
+		res.Warning = "datasource plugin not found"
+	} else if ref.Type == "" {
+		ref.Type = ds.Type // awkward! but makes the reporting accurate for dashboards before schemaVersion 36
```

Comment:
shouldn't this be done at the call site of the `Resolve` function? not a fun of mutating input args - it can lead to tricky bugs

Reply:
yes -- this is pretty messy and specifically around old dashboard>datasource references.

We could skip it entirely, but then would not have an accurate/easy way to run stats on which datasource types are referenced.  Maybe supporting that is not worth the mutating 🤷

Reply:
can't we have both if we just return the `key` _and_ `type`?